dnf5/commands/list/list.cpp

```diff
@@ -0,0 +1,284 @@
+}
+
+std::unique_ptr<libdnf::cli::output::PackageListSections> ListCommand::create_output() {
+    auto out = std::make_unique<libdnf::cli::output::PackageListSections>();
+    out->setup_cols();
+    return out;
+}
+
+void ListCommand::run() {
+    auto & ctx = get_context();
+    auto & config = ctx.base.get_config();
+    auto showduplicates = config.showdupesfromrepos().get_value();
+
+    libdnf::rpm::PackageQuery full_package_query(ctx.base);
+    libdnf::rpm::PackageQuery base_query(ctx.base);
+
+    // pre-select by patterns
+    if (!pkg_specs.empty()) {
+        base_query = libdnf::rpm::PackageQuery(ctx.base, libdnf::sack::ExcludeFlags::APPLY_EXCLUDES, true);
+        libdnf::ResolveSpecSettings settings{.with_nevra = true, .with_provides = false, .with_filenames = false};
+        for (const auto & spec : pkg_specs) {
+            libdnf::rpm::PackageQuery pkg_query(full_package_query);
+            pkg_query.resolve_pkg_spec(spec, settings, true);
+            base_query |= pkg_query;
+        }
+    }
+
+    // did any package match the criteria?
+    bool package_matched = false;
+    // output table
+    auto sections = create_output();
+
+    switch (pkg_narrow) {
+        case PkgNarrow::ALL: {
+            libdnf::rpm::PackageQuery installed(base_query);
+            installed.filter_installed();
+            libdnf::rpm::PackageQuery available(base_query);
+            available.filter_available();
+            if (!showduplicates) {
+                available.filter_priority();
+                available.filter_latest_evr();
+                // keep only those available packages that are either not installed or
+                // available EVR is higher than the installed one
+                libdnf::rpm::PackageQuery installed_latest(installed);
+                installed_latest.filter_latest_evr();
+                available.filter_nevra(installed_latest, libdnf::sack::QueryCmp::NOT | libdnf::sack::QueryCmp::LTE);
+            }
+            package_matched |= sections->add_section("Installed packages", installed);
+            {
+                // TODO(mblaha) currently only the installed version and upgrades
+                // are highlighted to make the output a bit saner
+                auto colorizer = std::make_unique<libdnf::cli::output::PkgColorizer>(
+                    installed,
+                    "",  //config.color_list_available_install().get_value(),
+                    "",  //config.color_list_available_downgrade().get_value(),
+                    config.color_list_available_reinstall().get_value(),
+                    config.color_list_available_upgrade().get_value());
+                package_matched |= sections->add_section("Available packages", available, colorizer);
+            }
+            break;
+        }
+        case PkgNarrow::INSTALLED: {
+            base_query.filter_installed();
+            package_matched |= sections->add_section("Installed packages", base_query);
```

Comment:
I would find it more consistent if we colored the installed and upgradable packages even if their option is used and they are the only packages in output.

Reply:
Fixed.